DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1

```diff
@@ -408,181 +552,62 @@ function Test-RuleProperties
 
     $desiredConfigurationMatch = $true
 
-    if ($Name -and ($FirewallRule.Name -ne $Name))
+    # Loop throug the ParameterList array and compare the source
+    # with the value of each parameter. If different then
+    # set $desiredConfigurationMatch to false.
+    foreach ($p in $ParameterList)
     {
-        Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
-            $($LocalizedData.PropertyNoMatchMessage) -f 'Name',$FirewallRule.Name,$Name
-            ) -join '')
-        $desiredConfigurationMatch = $false
-    }
-
-    if ($DisplayName -and ($FirewallRule.DisplayName -ne $DisplayName))
-    {
-        Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
-            $($LocalizedData.PropertyNoMatchMessage) -f 'DisplayName',$FirewallRule.DisplayName,$DisplayName
-            ) -join '')
-        $desiredConfigurationMatch = $false
-    }
-
-    if ($Group -and ($FirewallRule.Group -ne $Group))
-    {
-        Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
-            $($LocalizedData.PropertyNoMatchMessage) -f 'Group',$FirewallRule.Group,$Group
-            ) -join '')
-        $desiredConfigurationMatch = $false
-    }
-
-    if ($Enabled -and ($FirewallRule.Enabled.ToString() -ne $Enabled))
-    {
-        Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
-            $($LocalizedData.PropertyNoMatchMessage) -f 'Enabled',$FirewallRule.Enabled.ToString(),$Enabled
-            ) -join '')
-        $desiredConfigurationMatch = $false
-    }
-
-    if ($Action -and ($FirewallRule.Action -ne $Action))
-    {
-        Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
-            $($LocalizedData.PropertyNoMatchMessage) -f 'Action',$FirewallRule.Action,$Action
-            ) -join '')
-        $desiredConfigurationMatch = $false
-    }
-
-    if ($Profile)
-    {
-        [String[]] $networkProfileinRule = $FirewallRule.Profile.ToString() -replace(' ', '') -split(',')
-
-        if ($networkProfileinRule.Count -eq $Profile.Count)
+        $ParameterSource = (Invoke-Expression -Command "`$($($p.source))")
+        $ParameterNew = (Invoke-Expression -Command "`$$($p.name)")
+        switch ($p.type)
         {
-            foreach($networkProfile in $Profile)
-            {
-                if (-not ($networkProfileinRule -contains $networkProfile))
+            'String' {
+                # Perform a plain string comparison.
+                if ($ParameterNew -and ($ParameterSource -ne $ParameterNew))
                 {
                     Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
                         $($LocalizedData.PropertyNoMatchMessage) `
-                            -f 'Profile',($networkProfileinRule -join ','),($Profile -join ',')
+                            -f $p.Name,$ParameterSource,$ParameterNew
                         ) -join '')
                     $desiredConfigurationMatch = $false
```

Comment:
Given at this point we have our tests down to two places where we can hit a false would it just make sense to just return $false here instead of assigning the variable? It means we should exit out of the code sooner instead of continuing to iterate through.

Also means the logic comes down to
````powershell
'String'
{
   # Parameter doesn't match
return $false
}
'Array'
{
   # do array compare
  return $false
}
````